lib/Platform/Intl/BCP47Parser.cpp

```diff
@@ -0,0 +1,539 @@
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alpha{4};
+  return isCharType(subtagRef, 4, 4, &isASCIILetter);
+}
+bool isUnicodeRegionSubtag(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (alpha{2} | digit{3});
+  return isCharType(subtagRef, 2, 2, &isASCIILetter) ||
+    isCharType(subtagRef, 3, 3, &isASCIIDigit);
+}
+bool isUnicodeVariantSubtag(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (alphanum{5,8} | digit alphanum{3});
+  return isCharType(subtagRef, 5, 8, &isASCIILetterOrDigit) ||
+    isCharType(subtagRef, 3, 3, &isASCIILetterOrDigit);
+}
+bool isUnicodeExtensionAttribute(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alphanum{3,8};
+  return isCharType(subtagRef, 3, 8, &isASCIILetterOrDigit);
+}
+bool isUnicodeExtensionKey(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alphanum alpha;
+  if (subtagRef.size() != 2) {
+    return false;
+  }
+  return isASCIILetterOrDigit(subtagRef.front()) && isASCIILetter(subtagRef.back());
+}
+bool isUnicodeExtensionKeyTypeItem(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alphanum{3,8};
+  return isCharType(subtagRef, 3, 8, &isASCIILetterOrDigit);
+}
+bool isTransformedExtensionKey(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alpha digit;
+  if (subtagRef.size() != 2) {
+    return false;
+  }
+  return isASCIILetter(subtagRef.front()) && isASCIIDigit(subtagRef.back());
+}
+bool isTransformedExtensionTValueItem(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (sep alphanum{3,8})+;
+  return isCharType(subtagRef, 3, 8, &isASCIILetterOrDigit);
+}
+bool isPrivateUseExtension(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (sep alphanum{1,8})+;
+  return isCharType(subtagRef, 1, 8, &isASCIILetterOrDigit);
+}
+bool isOtherExtension(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (sep alphanum{2,8})+;
+  return isCharType(subtagRef, 2, 8, &isASCIILetterOrDigit);
+}
+}
+
+class LanguageTagParser {
+ public:
+  LanguageTagParser(const std::u16string &localeId);
+  LanguageTagParser();
+  ~LanguageTagParser();
+
+  // public function declaration
+  bool parseUnicodeLocaleId();
+  ParsedLocaleIdentifier getParsedLocaleId();
+  std::u16string toString();
```

Comment:
Is this ever used?

Reply:
No, I used to use it for testing, but I don't think it is necessary.